designs/disruption-controls-by-method.md

```diff
@@ -0,0 +1,469 @@
+* 👎👎 Adds complexity to understanding which disruptionMethod is associated with a particular disruption reason. 
+* 👎 Makes validation of a particular disruption 
+### Approach B: Defining Per Method Controls  
+Ideally, we could move all generic controls that easily map into other actions into one set of action controls, this applies to budgets and other various disruption controls that could be more generic. 
+### Proposed Spec 
+```go
+type Disruption struct {
+    All		  DisruptionSpec `json:defaults"`
+    Consolidation ConsolidationSpec `json:"consolidation"`
+    Drift         DriftSpec         `json:"drift"`
+    Expiration    ExpirationSpec    `json:"expiration"`
+    Emptiness     EmptinessSpec     `json:"emptiness"`
+}
+
+type DisruptionCommonSpec struct {
+    DisruptAfter string   `json:"disruptAfter"`
+    Budgets      []Budget `json:"budgets"`
+}
+
+type ConsolidationSpec struct {
+    DisruptionCommonSpec
+    ConsolidationPolicy string `json:"consolidationPolicy"`
+}
+
+type DriftSpec struct {
+    DisruptionCommonSpec
+}
+
+type ExpirationSpec struct {
+    DisruptionCommonSpec
+}
+
+type EmptinessSpec struct {
+    DisruptionCommonSpec
+}
+
+type Budget struct {
+    Nodes    string  `json:"nodes"`
+    Schedule *string `json:"schedule,omitempty"`
+    Duration *string `json:"duration,omitempty"`
+    Reason string 
+}
+}
+
+type DisruptionMethod string 
+
+const (
+	All DisruptionMethod = ""
+	Consolidation DisruptionMethod = "Consolidation" 
+	Drift DisruptionMethod = "Drift" 
+	Expiration DisruptionMethod = "Expiration" 
+	Emptiness DisruptionMethod = "Emptiness"
+)
+```
+#### Example 
+
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: example-nodepool
+spec:
+  disruption:
+    defaults:
+      budgets: 
+        - nodes: 10% 
+          schedule: "0 0 1 * *"
+          duration: 1h 
+    consolidation:
+      consolidationPolicy: WhenUnderutilized
+      disruptAfter: "30m"
+    drift:
+      budgets:
+        - nodes: "20%"
+          schedule: "0 0 1 * *"
+          duration: "1h"
+          drift:
+            disruptAfter: "1h"
+            budgets:
+              - nodes: "10%"
+                reason: "NodeImageDrift"
+                schedule: "0 0 * * 0"
+                duration: "2h"
+              - nodes: "50%" 
+                reason: "K8sVersionUpgrade"
+                schedule: "@yearly"
+    expiration:
+      disruptAfter: "Never"
+```
+
+#### Reasons
+In this design, rather than methods being defined at the budget level, we add an additonal layer of abstraction. Then for each budget, we apply a reason or All/Undefined. If reason isn't specifed in a budget we take the same behavior in terms of fallback for all on these method types.
+
+This design allows for simplification of reason as its very easy to directly define a relationship between a given disruption method and its sub-action since the disruption method is explicitly declared. 
```

Comment:
I am removing the full section I put on all the considerations of reason, because i think it complicates the design too much, I might add some sort of extension of this design for the reasons. For now I am just going to simplify things and not add too much abscraction. 


Im going to dump those notes here: 

-----------------------------------------------
### Reasons
### Q: Should Karpenter allow for more granular disruption reasons to have budgets? like for specific drift reasons for example?
Biggest user story here is users may want to handle particular drift reasons in different ways.

In the current setup, Karpenter provides a disruption controller with standard Method implementations. However, there's a need for more granularity in defining disruption budgets. For example, users may want to have a different schedule for Kubernetes version upgrades compared to node image upgrades which are both driven via cloudprovider.IsDrifted().
Karpenter should provide a way to extend these more granular reasons that are children of the disruption methods. It does add a signficant number of questions into the mix. But almost deserves its own design as it opens up a multitude of questions.  

The [AWS Provider](https://github.com/search?q=repo%3Aaws%2Fkarpenter-provider-aws+cloudprovider.DriftReason&type=code), [Azure Provider](https://github.com/search?q=repo%3AAzure%2Fkarpenter-provider-azure+cloudprovider.DriftReason&type=code), and [Core](https://github.com/search?q=repo%3Akubernetes-sigs%2Fkarpenter+DriftReason&type=code) have the following drift reasons.

**Core**
- NodePoolDrifted
- RequirementsDrifted

**AWS** 
- AMIDrift
- SubnetDrift 
- SecurityGroupDrift 
- NodeClassDrift

**Azure** 
- K8sVersionDrift 
- ImageVersionDrift

As you can see there are quite a few cases for drift, and the type of action that is taken from these forms of drift are very different. This leads to Drift Reasons needing a place in the api at some stage.

#### Q: If all the reasons mainly apply to drift, why add a Method paired with reason rather than only allowing DriftReason and specifying that alongside drift only?
Currently the method interface has a Type() implying that other methods also will want to specify disruption method types at a finer granularity 

```go
type Method interface {
	ShouldDisrupt(context.Context, *Candidate) bool
	ComputeCommand(context.Context, map[string]int, ...*Candidate) (Command, error)
	Type() string
	ConsolidationType() string
}
```



#### Q: Why have the distinction between method and reason? Why not just have everything be a reason?
The distinction between method and reason is crucial for providing both a high-level and a granular control over disruptions. The method corresponds to the type of disruption action, such as "Drift" or "Consolidation", which is a broad category of disruption. Within each method, there can be multiple reasons that provide specific context for the disruption, such as "AMIDrift" in the case of AWS.

By separating method and reason, Karpenter allows users to define budgets and policies at both levels. Users can set a general budget for all "Drift" disruptions, and then further refine the control by specifying different budgets or schedules for specific reasons like "AMIDrift". This two-tiered approach offers flexibility and precision, enabling users to manage disruptions more effectively according to their needs.

Moreover, the distinction helps in maintaining clarity and organization within the API. It allows for a structured way to handle disruptions, where methods can be seen as categories, and reasons as subcategories. This hierarchy makes it easier for users to navigate and understand the disruption policies they have set up.

This also allows karpenter to easily tell inside of the disruption controller which actions it needs to be looking for when checking Type(). Without this top method, it becomes much more challenging to match a Type() with a Reason. 


#### Q: Budgets currently work by tracking deletion in total. In this new system, karpenter has to be aware of each disruption method + reason, does adding two fields Method + Reason make it harder to track how much of budget we have used? Should we add DisruptionReason to the nodeclaim? 
To properly track nodeclaims in deleting state for each nodepool effectively and easily in cluster state adding an additional field or status condition to indicate why the nodeclaim is being disrupted/deleted  makes a lot of sense. A single reason makes it easier to track on the nodeclaim.
##### Q: Should we have two status conditions for disruption method and disruption reason? Or should they be consolidated into one condition?
Having two separate status conditions for disruption method and disruption reason could potentially provide more detailed information about the disruption. However, it might also complicate the tracking process. On the other hand, consolidating them into one condition would simplify the tracking but might lack some details. Considering the trade-off between detail and simplicity, it would be more practical to consolidate them into one condition. This way, we can track the disruption process more efficiently while still maintaining necessary information about the disruption.


This is separate of the concept of NodePool budgets Reason, and just refers to a general indication of a disruption reason we put on the nodeclaim to easily track disruption for each disruption method.

#### Q: For a conflicting child reason, do we respect the parent method? 
```yaml 
spec:
  disruption:
    budgets:
    - nodes: 50
      method: "Drift" 
      reason: "NodeImageDrift"
    - nodes: 25 
      method: "Drift" 

```
In this case, do we only allow for 25 disruptions for any drift method regardless of the reason?  Or Do we say that for NodeImageDrift, we allow 50 and for all other drift reasons we allow 25? I would say the ladder is the desirable behavior and easiest to reason about.

Reply:
Removing because "It seems that perfection is attained, not when there is nothing more to add, but when there is nothing more to take away"